php_thread.go

```diff
@@ -46,5 +99,55 @@ func (thread *phpThread) pinString(s string) *C.char {
 
 // C strings must be null-terminated
 func (thread *phpThread) pinCString(s string) *C.char {
-	return thread.pinString(s+"\x00")
+	return thread.pinString(s + "\x00")
+}
+
+//export go_frankenphp_before_script_execution
+func go_frankenphp_before_script_execution(threadIndex C.uintptr_t) *C.char {
+	// returning nil signals the thread to stop
+	if threadsAreDone.Load() {
+		return nil
+	}
+	thread := phpThreads[threadIndex]
+
+	// if the thread is not ready, set it up
+	if !thread.isReady.Load() {
+		thread.isReady.Store(true)
```

Comment:
My rule of thumb is that if you find yourself using atomics, it's probably really hard to get right (locks are usually better). There's a small race condition here where two concurrent threads can enter this if-statement.

If you want to continue to use atomics here, then you are probably looking at something like this:

```suggestion
	if thread.isReady.CompareAndSwap(false, true) {
```

which will only enter the block if it was set to `true`.

Reply:
I'll change it to a CompareAndSwap 👍 . The reason this currently isn't a race condition stems from the fact that `isReady` can only be set to true on the next line.
The original idea for this to be an atomic Bool instead of a regular one was to use it in metrics. But if you prefer pushing metrics instead of pulling them, this could also be made a regular bool.

Reply:
I've thought about having a go-routine every so often to gather metrics, but it would definitely violate every sensible coding methodology (SOLID, Clean Code, etc) in existence. Pushing them is not as pretty, but at least it is decoupled.

Reply:
This is now done with thread states